experimental/dds/tree2/src/shared-tree/treeCheckout.ts

```diff
@@ -272,7 +272,11 @@ export class TreeCheckout implements ITreeCheckoutFork {
 			HasListeners<CheckoutEvents>,
 		private readonly removedTrees: DetachedFieldIndex = makeDetachedFieldIndex("repair"),
 	) {
-		branch.on("change", (event) => {
+		// We subscribe to `beforeChange` rather than `afterChange` here because it's possible that the change is invalid WRT our forest.
+		// For example, a bug in the editor might produce a malformed change object and thus applying the change to the forest will throw an error.
+		// In such a case we will crash here, preventing the change from being added to the commit graph, and preventing `afterChange` from firing.
+		// One important consequence of this is that we will not submit the op containing the invalid change, since op submissions happens in response to `afterChange`.
+		branch.on("beforeChange", (event) => {
```

Comment:
Does applying the delta prior to it being present in the graph mean that any events that read the graph (e.g. when we expose branching shortly after MVP) will have an out-of-date graph?

Reply:
Good question. Clients can't read the graph directly via the public API (even after we expose branching). They can however interact with it indirectly via a fork, merge, or a rebase. They're all pretty sus to do during a change, though:

* If you fork in the middle of a change event, your fork will either "after all the changes in the new commit" (behavior before this PR) or "before all the changes in this commit" (behavior after this PR). Both are weird. If you fork when you've only been notified of half of the changes in a commit, you can't produce a branch that has only that half of the changes. You have to round up or down.
* Rebasing and merging are even more problematic because they might actually mutate the heads of the branch. If you rebase over another branch, or merge another branch into your branch during a change, then you're causing a change in the graph during a change in the graph - bad idea, just like changing the tree while changing the tree. You can technically have another branch rebase over _you_, without mutating yourself, and you can merge _into_ another branch, without mutating yourself... but both of those scenarios have the same "halfway done" weirdness like when making a fork, above.

I hadn't thought about this deeply - thanks for the prompt. We should probably ban all branch operations during changes, just like we will ban edits.